pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignmentStrategyFactory.java

```diff
@@ -0,0 +1,89 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.assignment.segment;
+
+import java.util.Map;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.config.table.assignment.SegmentAssignmentConfig;
+import org.apache.pinot.spi.utils.CommonConstants.Segment.AssignmentStrategy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Utility class for segment assignment strategy.
+ */
+public class SegmentAssignmentStrategyFactory {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentAssignmentStrategyFactory.class);
+  private SegmentAssignmentStrategyFactory() {
+  }
+
+  public static SegmentAssignmentStrategy
+  getSegmentAssignmentStrategy(TableConfig tableConfig, int numReplicaGroups, int numPartitions) {
+    String assignmentStrategy = null;
+    TableType currentTableType = tableConfig.getTableType();
+    // Handle current replica group segment assignment
+    if (tableConfig.getValidationConfig().getSegmentAssignmentStrategy() != null) {
+      assignmentStrategy = tableConfig.getValidationConfig().getSegmentAssignmentStrategy();
+    }
+    // Accommodate new changes for assignment strategy
+    Map<InstancePartitionsType, SegmentAssignmentConfig>
+        segmentAssignmentConfigMap = tableConfig.getSegmentAssignmentConfigMap();
+
+    if (tableConfig.isDimTable()) {
+      // Segment Assignment Strategy for DIM tables
+      return new DimTableSegmentAssignmentStrategy();
+    } else {
+      // Try to determine segment assignment strategy from table config
+      if (segmentAssignmentConfigMap != null) {
+        SegmentAssignmentConfig segmentAssignmentConfig;
+        // Use the pre defined segment assignment strategy
+        if (currentTableType == TableType.OFFLINE) {
+          segmentAssignmentConfig = segmentAssignmentConfigMap.get(InstancePartitionsType.OFFLINE);
+        } else {
+          segmentAssignmentConfig = segmentAssignmentConfigMap.get(InstancePartitionsType.COMPLETED);
+        }
+        // segment assignment config is only applicable to offline tables and completed segments of real time tables
+        if (segmentAssignmentConfig != null) {
+          assignmentStrategy = segmentAssignmentConfig.getAssignmentStrategy();
+        }
+      }
+    }
+    // use the existing information to determine segment assignment strategy
+    if (assignmentStrategy == null) {
+      if (numReplicaGroups == 1 && numPartitions == 1) {
+        return new BalanceNumSegmentAssignmentStrategy();
```

Comment:
Should not directly return. Let's first assign this to the variable. We would like to override the value if we configure the custom assignment in the table config.

Reply:
updated